### PR TITLE
Fix FBGEMM build with gcc 9.3+

### PR DESCRIPTION
**Pull request by ykim362**
### Description
Build failed with gcc 9.3+ when USE_FBGEMM is on.

This PR fixes bugs reported #835 and #709 .

List of changes:
- In top level CMakeList.txt, '-march=native' is turned off when built with FBGEMM.

Added dependencies: none

### How to test
Compile with gcc 9.3.

### Checklist

- [x] I have tested the code manually
- [ ] I have run regression tests
- [x] I have read and followed CONTRIBUTING.md
- [x] I have updated CHANGELOG.md

### CODE REVIEW

**Comment by emjotde** [comment]
Hm, that's a very broad change to fix such a specific issue. Can we narrow this down a bit? Do we know what exactly is going on here?

**Comment by ykim362** [comment]
@emjotde When compiling avx2 target on avx512 machines, the newer compilers are doing some target specific optimizations which generates broken assembly. This only affects when USE_FBGEMM is on though. The best thing we can do is removing '-march=native' for avx2 target on avx512 machines. But, it's not likely we are getting too much from the native architecture flag.

**Comment by ykim362** [comment]
@emjotde I made a change that the original CMAKE_CXX_FLAGS is not touched.

**Comment by snukky** [comment]
I updated the branch with master to fix failures in GitHub checks that were unrelated to this PR.

File: CMakeLists.txt [hunk 1]
```diff
@@ -216,6 +216,11 @@ else(MSVC)
   set(CMAKE_C_FLAGS_PROFILE         "${CMAKE_C_FLAGS_RELEASE} -pg")
   set(CMAKE_C_FLAGS_PROFGEN         "${CMAKE_C_FLAGS_RELEASE} -fprofile-generate -fprofile-correction")
   set(CMAKE_C_FLAGS_PROFUSE         "${CMAKE_C_FLAGS_RELEASE} -fprofile-use -fprofile-correction")
+
+  # '-march=native' makes avx2 assembly generation fail on AVX512 CPUs with gcc 9.3+
+  if (USE_FBGEMM)
+    string(REPLACE "-march=native" "" CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS}")
```
Comment (snukky) [inline CMakeLists.txt:222]: Maybe we should apply this only if GCC is >= 9.2? For example (I did not test it):

    if(USE_FBGEMM AND CMAKE_COMPILER_IS_GNUCC AND CMAKE_CXX_COMPILER_VERSION VERSION_GREATER 9.2)

**Comment by emjotde** [comment]
That is OK, I think. Please add a message that gets displayed when GCC 9.3+ is used.

**Comment by ykim362** [comment]
@emjotde I've added a message.

**Comment by kpu** [comment]
All the more reason to dis FBGEMM but really scope this stuff instead of breaking it for everything.

**Comment by emjotde** [comment]
@kpu Sorry, can be a bit clearer regarding the second part?

**Comment by kpu** [comment]
This isn't really about `-march=native`, it's about the flags that it happens to map to on some hardware.  And should be able to compile Marian with whatever architecture flags, such as `-mavx512bw`, I want. If a specific file has a problem with that, unset the flags only for that file, not the entire project.

**Comment by emjotde** [comment]
Happy to comply. Feel free to use the review interface to leave a hint where and how to do that.

**Comment by kpu** [comment]
Go fix your cmake around her to filter options: https://github.com/marian-nmt/FBGEMM/blob/64888200cbc9e9c287601f23c58b4088cdf51cba/CMakeLists.txt#L126

**Comment by emjotde** [comment]
@ykim362 can you take a look? That seems indeed like a better idea.

**Comment by ykim362** [comment]
@emjotde @kpu Sounds good. I will try to fix there, and will close this PR if that works. Thanks!

**Comment by ykim362** [comment]
@emjotde @kpu Now, all the changes are in the FBGEMM submodule. It works fine. Please take a look. Thanks!

**Comment by snukky** [comment]
Could you also mention this fix in the changelog before merging it into master? Thanks!

**Comment by ykim362** [comment]
@snukky I added the fix in the changelog. Thanks!

**Comment by emjotde** [comment]
Thanks @ykim362

**Review by emjotde — APPROVED** [review]

**Pull request merged**